### ISSUE TITLE
Duplicates in categorical search space

### ISSUE THREAD

**Issue by clementruhm** [google/vizier#1204]
Hi @xingyousong , kicking off new issue with the example to reproduce. Will close the [previous one](https://github.com/google/vizier/issues/1199).

I have several discrete / categorical / boolean parameters as a search space. There is a validator which checks if sampled config is valid or not. If not, configuration is marked as infeasible. Finally I need to sample a batch (population) of suggestions to compute metrics in parallel. Here is the code:

```python
import time
import random
import logging

from vizier import algorithms as vza
from vizier.algorithms import designers
from vizier.service import pyvizier as vz


search_space = {
    'context': [True, False],
    'blocks': [True, False],
    'policy': ['g', 'u'],
    'dtype': ['fp16', 'bp16'],
    'mode': ['disable', 'auto', 'fp16', 'fp8'],
    'bs': [64, 128, 256, 512, 1024, 2048, 4098],
    'tokens': [512, 1024, 2048, 4096, 8192, 12288, 20480, 28672, 36864],
    'multiple': [True, False],
    'stride': [32, 64, 128, 256, 512],
    'extra': [True, False],
    'use_fuse': [False, True],
    'use_page': [True, False],
}


def is_valid(values) -> bool:
    """
    set of rules that check if the sampled hyper parameter combination is valid
    """
    if values["dtype"] == "fp16" and values["mode"] == "fp8":
        return False
    if values["bs"] >  values["tokens"]:
        return False
    if values["stride"] > values["bs"]:
        return False
    if values["extra"] and values["multiple"]:
        return False
    if values["use_fuse"] and not values["blocks"]:
        return False
    if values["use_page"] and values["policy"] == "g":
        return False
    return True


def create_designer():
    problem = vz.ProblemStatement()
    for name, vals in search_space.items():
        if type(vals[0]) == int or type(vals[0]) == float:
            problem.search_space.root.add_discrete_param(name, feasible_values=vals)
        elif type(vals[0]) == bool:
            problem.search_space.root.add_bool_param(name)
        else:
            # str or enum
            problem.search_space.root.add_categorical_param(name, feasible_values=vals)

    goal = vz.ObjectiveMetricGoal.MAXIMIZE
    metric_info = vz.MetricInformation(name="target_metric", goal=goal)
    problem.metric_information.append(metric_info)

    designer = designers.VizierGPUCBPEBandit(problem)
    return designer


BATCH_SIZE = 4
POPULATION_NUM = 2
POPULATION_SIZE = 4


def main():
    designer = create_designer()

    missing_param = 0
    total_suggestions = 0
    duplicate_suggestions = 0
    pending_suggestions = 0
    invalid_config = 0

    previous_suggestions = set()

    # I need POPULATION_NUM populations, each having POPULATION_SIZE samples
    for i in range(POPULATION_NUM):

        population = []
        pending = set()
        start = time.time()
        while len(population) < POPULATION_SIZE:
            suggestions = designer.suggest(count=BATCH_SIZE)
            total_suggestions += 1
            suggestion = random.choice(suggestions)

            if not all(x in suggestion.parameters for x in set(search_space.keys())):
                logging.warning("Parameter is missing from the suggestion")
                missing_param += 1
                continue

            suggestion_dict = {}
            for name in search_space.keys():
                val = suggestion.parameters[name].value
                # convert the type (why we have to do it?)
                if type(search_space[name][0]) == int:
                    val = int(val)
                elif type(search_space[name][0]) == bool:
                    val = val.lower() == "true"
                suggestion_dict[name] = val
            
            suggestion_id = hash(str(suggestion_dict))
            if suggestion_id in previous_suggestions:
                if suggestion_id in pending:
                    logging.warning("Pending suggestion")
                    pending_suggestions += 1
                else:
                    logging.warning("Duplicate suggestion")
                    duplicate_suggestions += 1
                continue
            previous_suggestions.add(suggestion_id)
            
            if not is_valid(suggestion_dict):
                trial = suggestion.to_trial()
                trial.complete(vz.Measurement(), infeasibility_reason="invalid_config")
                designer.update(vza.CompletedTrials([trial]), vza.ActiveTrials())
                logging.warning("Invalid combination")
                invalid_config += 1
                continue
            population.append(suggestion)
            pending.add(suggestion_id)
        
        took = time.time() - start
        logging.info(f"Got population #{i}. It took {took} seconds")
        
        # report back some random values for the population
        for suggestion in population:
            result = {"target_metric": random.uniform(1000, 10000)}
            trial = suggestion.to_trial()
            trial.complete(vz.Measurement(metrics=result))
            designer.update(vza.CompletedTrials([trial]), vza.ActiveTrials())

        logging.info(f"After population {i}, sampled {total_suggestions} suggestions")
        logging.info(f"{missing_param} were missing a param")
        logging.info(f"{invalid_config} were invalid")
        logging.info(f"{duplicate_suggestions} were duplicates")
        logging.info(f"{pending_suggestions} were suggested again")


if name == "main":
    logging.basicConfig(level=logging.INFO)
    main()
```

When running this example there are following issues:

1) during process of sampling 2 populations with 4 suggestions each, I get +100 duplicates (see how duplicates are computed). Is there a way to avoid it? in case of larger populations (for ex. 8 suggestions) there are more duplicates.
2) it takes *significant time to get a population*, 0.5 hours for this example. This is due to lengthy sampling process. I tried to run this on GPU but I can't see significant speed ups. Is there a way to run this example faster? maybe some problem statement or designer configuration that I should tweak?
3) In case of larger population, sometimes designer provides a suggestion, where not all the parameters are provided. I.e. `missing_param` in the code. Happens rather rarely.
4) If I request just one suggestion `designer.suggest(count=1)`, then I always get duplicates. I think count=1 corresponds to UCB-only (section 3.6, eq. 5 from the [paper](https://arxiv.org/pdf/2408.11527), count=4 allows to trigger PE (eq. 7) and get some randomized configurations. Number of duplicates is still pretty big, so process is slow.
5) the script consumes quite a bit of RAM. up to 10gb easily. memory consumption grows throughout the optimization process. How much memory the optimization may potentially take? is there an upper bound? can I impose a limit? also I can see that JAX stores a lot of intermediate files, that clog the disk. 
6) designer messes the return types a bit. hence the conversion: `val = val.lower() == "true"`. Can designer return python types?
7) The dependency management seems to be a bit off. for the fresh environment:
```bash
conda create -n test python=3.10
conda activate test
pip install google-vizier[algorithms,jax]
```

I am getting:
```bash
File "/home/user/miniconda3/envs/test/lib/python3.10/site-packages/cvxpy/__init__.py", line 18, in <module>
    import cvxpy.interface.scipy_wrapper
  File "/home/user/miniconda3/envs/test/lib/python3.10/site-packages/cvxpy/interface/scipy_wrapper.py", line 37, in <module>
    method = getattr(spmatrix, method_name)
AttributeError: type object 'spmatrix' has no attribute 'div'. Did you mean: 'dir'?
```

unless i manually downgrade scikit learn and scipy:
```
pip install scikit-learn==1.1.2 scipy==1.10.1
```

should the versions be frozen?

8) it all works only for python3.10. Is there chance to support newer python, for ex. 3.12? what are the blockers for that?

Hope that helps to narrow down. Maybe i am not using the API correctly, but I mostly followed the examples from the docs. Looking forward for the insights.

Kind regards

**Comment by xingyousong** [comment]
Got it - TL;DR Please **never** use the low-level designer API unless you absolutely know what you're doing.

Unless you're trying to tweak the algorithm behavior, it's absolutely recommended to use the [client-side basic user API](https://oss-vizier.readthedocs.io/en/latest/guides/user/running_vizier.html). This is because the client API has done all the heavy lifting of [wrapping the designer and dealing with all the edge-cases and intended use-cases](https://github.com/google/vizier/blob/main/vizier/_src/algorithms/policies/designer_policy.py).

To parallelize your evaluation workload, instead of writing this complicated population logic, you can simply call multiple clients in a thread-pool, all of whom are simply performing a suggest-then-evaluate loop.

Specific issues:
* **Duplicated trials:** Your code is misusing `designer.update()` by not providing active trials (i.e. all unevaluated suggestions that have been produced so far), which explains the duplicate behavior. Note that the algorithm explicitly needs to know what are the "active trials" in order to de-duplicate (see how [we use it in the service](https://github.com/google/vizier/blob/main/vizier/_src/algorithms/policies/designer_policy.py#L96-L103)).
* **Converting output types:** Use `add_int_param(...)` instead of `add_discrete_param` for your integer parameters, so that the conversions are handled automatically. I also think that we added "external type" conversion in the most recent pip installation, so try installing the latest version of Vizier.
* **Installation:** Only do `pip install google-vizier[jax]` which installs the Vizier GP variants. `algorithms` as an option contains many more algorithms which you'll never need.

**Comment by clementruhm** [comment]
the problem is that I dont really own the evaluation in my code. I prepare a population of configurations, send them out, then get back results in async manner in no particular order. The documentation is [not very verbose](https://oss-vizier.readthedocs.io/en/latest/guides/user/running_vizier.html#client-parallelization) about multiple clients. I found this [tutorial](https://colab.research.google.com/github/google/vizier/blob/main/docs/tutorials/automl_conf_2023.ipynb#scrollTo=F5e2FAhCb5pc) and tried to refactor my snippet to use client API to at least sample configs from multiple clients in a threadpool:

```python
import time
import random
import logging
import threading
from multiprocessing.pool import ThreadPool

from vizier import service
from vizier.service import clients
from vizier.service import pyvizier as vz


search_space = {
    'context': [True, False],
    'blocks': [True, False],
    'policy': ['g', 'u'],
    'dtype': ['fp16', 'bp16'],
    'mode': ['disable', 'auto', 'fp16', 'fp8'],
    'bs': [64, 128, 256, 512, 1024, 2048, 4098],
    'tokens': [512, 1024, 2048, 4096, 8192, 12288, 20480, 28672, 36864],
    'multiple': [True, False],
    'stride': [32, 64, 128, 256, 512],
    'extra': [True, False],
    'use_fuse': [False, True],
    'use_page': [True, False],
}


def is_valid(values) -> bool:
    """
    set of rules that check if the sampled hyper parameter combination is valid
    """
    if values["dtype"] == "fp16" and values["mode"] == "fp8":
        return False
    if values["bs"] >  values["tokens"]:
        return False
    if values["stride"] > values["bs"]:
        return False
    if values["extra"] and values["multiple"]:
        return False
    if values["use_fuse"] and not values["blocks"]:
        return False
    if values["use_page"] and values["policy"] == "g":
        return False
    return True


def create_problem():
    problem = vz.ProblemStatement()
    for name, vals in search_space.items():
        if type(vals[0]) == int or type(vals[0]) == float:
            problem.search_space.root.add_discrete_param(name, feasible_values=vals)
        elif type(vals[0]) == bool:
            problem.search_space.root.add_bool_param(name)
        else:
            # str or enum
            problem.search_space.root.add_categorical_param(name, feasible_values=vals)

    goal = vz.ObjectiveMetricGoal.MAXIMIZE
    metric_info = vz.MetricInformation(name="target_metric", goal=goal)
    problem.metric_information.append(metric_info)
    service.VIZIER_DB_PATH = "./to_remove.db"
    return problem


POPULATION_NUM = 2
POPULATION_SIZE = 4


def get_suggested_params(suggestion):
    suggestion_dict = {}
    for name in search_space.keys():
        val = suggestion.parameters[name]
        # still had to convert the type
        if type(search_space[name][0]) == int:
            val = int(val)
        elif type(search_space[name][0]) == bool:
            val = val.lower() == "true"
        suggestion_dict[name] = val
    return suggestion_dict 


missing_param = 0
invalid_config = 0
lock = threading.Lock()


def main():

    problem = create_problem()
    study_config = vz.StudyConfig.from_problem(problem)
    study_config.algorithm = 'DEFAULT'
    client = clients.Study.from_study_config(study_config, owner='me', study_id='example_study')

    previous_suggestions = set()
    total_suggestions = 0
    duplicate_suggestions = 0

    # I need POPULATION_NUM populations, each having POPULATION_SIZE samples
    for i in range(POPULATION_NUM):

        def sample_in_thread(client_id: int):
            global invalid_config, missing_param
            thread_client = clients.Study.from_resource_name(client.resource_name)
            while True:  # limit by certain number of attempts
                suggestion = thread_client.suggest(count=1, client_id=str(client_id))[0]

                if not all(x in suggestion.parameters for x in set(search_space.keys())):
                    logging.warning("Parameter is missing from the suggestion")
                    suggestion.complete(vz.Measurement(), infeasible_reason="incomplete_config")
                    with lock:
                        missing_param += 1  # make atomic
                    continue

                suggestion_dict = get_suggested_params(suggestion)
                if not is_valid(suggestion_dict):
                    suggestion.complete(vz.Measurement(), infeasible_reason="invalid_config")
                    logging.warning("Invalid params combination")
                    with lock:
                        invalid_config += 1  # make atomic
                    continue
                
                return suggestion
        
        start = time.time()
        pool = ThreadPool(POPULATION_SIZE)
        population = pool.map(sample_in_thread, range(POPULATION_SIZE))
        total_suggestions += len(population)
        took = time.time() - start
        logging.info(f"Got population #{i}. It took {took} seconds")

        # check if there are not duplicates
        unique_population = []
        for s in population:
            s_dict = get_suggested_params(s)
            s_id = hash(str(s_dict))
            if s_id in previous_suggestions:
                logging.error("Suggested a duplicate!!")
                duplicate_suggestions += 1
                continue
            previous_suggestions.add(s_id)
            unique_population.append(s)
        
        # complete the population
        for s in unique_population:
            result = {"target_metric": random.uniform(1000, 10000)}
            s.complete(vz.Measurement(metrics=result))

        logging.info(f"After population {i}, sampled {total_suggestions} suggestions")
        logging.info(f"{missing_param} were missing a param")
        logging.info(f"{invalid_config} were invalid")
        logging.info(f"{duplicate_suggestions} were duplicates")


if name == "main":
    logging.basicConfig(level=logging.INFO)
    main()
```  

but still I see in a very first loop iteration, the threadpool returns duplicated suggestions. 

I think the best way for me would be to use actually `suggest(count=population_size)`. But that one also produces duplicates :/

**Comment by clementruhm** [comment]
Here is example that uses `suggest(count=population_size)`. on the very first suggestion iteration sometimes it suggests exactly the same configs :(

```python
import time
import random
import logging
import threading
from multiprocessing.pool import ThreadPool

from vizier import service
from vizier.service import clients
from vizier.service import pyvizier as vz


search_space = {
    'context': [True, False],
    'blocks': [True, False],
    'policy': ['g', 'u'],
    'dtype': ['fp16', 'bp16'],
    'mode': ['disable', 'auto', 'fp16', 'fp8'],
    'bs': [64, 128, 256, 512, 1024, 2048, 4098],
    'tokens': [512, 1024, 2048, 4096, 8192, 12288, 20480, 28672, 36864],
    'multiple': [True, False],
    'stride': [32, 64, 128, 256, 512],
    'extra': [True, False],
    'use_fuse': [False, True],
    'use_page': [True, False],
}


def is_valid(values) -> bool:
    """
    set of rules that check if the sampled hyper parameter combination is valid
    """
    if values["dtype"] == "fp16" and values["mode"] == "fp8":
        return False
    if values["bs"] >  values["tokens"]:
        return False
    if values["stride"] > values["bs"]:
        return False
    if values["extra"] and values["multiple"]:
        return False
    if values["use_fuse"] and not values["blocks"]:
        return False
    if values["use_page"] and values["policy"] == "g":
        return False
    return True


def create_problem():
    problem = vz.ProblemStatement()
    for name, vals in search_space.items():
        if type(vals[0]) == int or type(vals[0]) == float:
            problem.search_space.root.add_discrete_param(name, feasible_values=vals)
        elif type(vals[0]) == bool:
            problem.search_space.root.add_bool_param(name)
        else:
            # str or enum
            problem.search_space.root.add_categorical_param(name, feasible_values=vals)

    goal = vz.ObjectiveMetricGoal.MAXIMIZE
    metric_info = vz.MetricInformation(name="target_metric", goal=goal)
    problem.metric_information.append(metric_info)
    return problem


POPULATION_NUM = 2
POPULATION_SIZE = 2


def get_suggested_params(suggestion):
    suggestion_dict = {}
    for name in search_space.keys():
        val = suggestion.parameters[name]
        # still had to convert the type
        if type(search_space[name][0]) == int:
            val = int(val)
        elif type(search_space[name][0]) == bool:
            val = val.lower() == "true"
        suggestion_dict[name] = val
    return suggestion_dict 


def main():

    problem = create_problem()
    study_config = vz.StudyConfig.from_problem(problem)
    study_config.algorithm = 'DEFAULT'
    client = clients.Study.from_study_config(study_config, owner='me', study_id='example_study')

    previous_suggestions = set()
    total_suggestions = 0
    duplicate_suggestions = 0
    missing_param = 0
    invalid_config = 0


    # I need POPULATION_NUM populations, each having POPULATION_SIZE samples
    for i in range(POPULATION_NUM):

        population = []
        start = time.time()
        while len(population) < POPULATION_SIZE:
            required = POPULATION_SIZE - len(population)
            suggestions = client.suggest(count=required)
            logging.info(f"Sampled {len(suggestions)} suggestions to complete {len(population)} / {POPULATION_SIZE}")
            total_suggestions += len(suggestions)

            # check if sampled suggestions are complete
            complete_suggestions = []
            for s in suggestions:
                if not all(x in s.parameters for x in set(search_space.keys())):
                    logging.warning("Parameter is missing from the suggestion")
                    s.complete(vz.Measurement(), infeasible_reason="incomplete_config")
                    missing_param += 1
                    continue
                complete_suggestions.append(s)

            # check if suggestions are valid
            valid_suggestions = []
            for s in complete_suggestions:
                s_dict = get_suggested_params(s)
                if not is_valid(s_dict):
                    s.complete(vz.Measurement(), infeasible_reason="invalid_config")
                    logging.warning("Invalid params combination")
                    invalid_config += 1  # make atomic
                    continue
                valid_suggestions.append(s)
            
            # add suggestions to population, checking that they are unique
            for s in valid_suggestions:
                s_dict = get_suggested_params(s)
                s_id = hash(str(s_dict))
                if s_id in previous_suggestions:
                    logging.error("Suggested a duplicate!!")
                    duplicate_suggestions += 1
                    continue
                previous_suggestions.add(s_id)
                population.append(s)

        took = time.time() - start
        logging.info(f"Got population #{i}. It took {took} seconds")

        # complete the population
        for s in population:
            result = {"target_metric": random.uniform(1000, 10000)}
            s.complete(vz.Measurement(metrics=result))

        logging.info(f"After population {i}, sampled {total_suggestions} suggestions")
        logging.info(f"{missing_param} were missing a param")
        logging.info(f"{invalid_config} were invalid")
        logging.info(f"{duplicate_suggestions} were duplicates")


if name == "main":
    logging.basicConfig(level=logging.INFO)
    main()
```

you mentioned that benchmark against other BO libs showed that vizier is superior. I confirm that Optuna also produces duplicates. there is a long discussion on the topic: https://github.com/optuna/optuna/issues/2021. TLDR: they say its not striaghtforward to fix from algorithmic perspective and what you need to do is to track the suggestions and sample again if there is a duplicate.

you explained that it would be and API abuse for vizier. can you think of another approach I can try to get batch of unique suggestions, eval them async and report back the results

**Comment by xingyousong** [comment]
I don't understand the use case here - 

I tried the batch setting with a more simple setting:

```
    for i in range(50):
      suggestions = client.suggest(count=5)
   
      # complete the population
      for s in suggestions:
          result = {"target_metric": random.uniform(1000, 10000)}
          s.complete(vz.Measurement(metrics=result))
```

And didn't find issues with duplicate suggestions, nor did I find any issues with missing parameters.
There's something very strange about the way you're performing the loop.

Can you verify this minimal case works fine for you?

**Comment by clementruhm** [comment]
That's fair. Running a minimal case produces duplicates much more rarely. For the search space from above, i used:

```python
for i in range(8):
    population = client.suggest(count=16)

    # check for duplicates

    for s in suggestions:
        result = {"target_metric": random.uniform(1000, 10000)}
        s.complete(vz.Measurement(metrics=result))

```

thats 128 trials in total from 330k search space.

Paper suggests that discrete params are rounded to the nearest feasible value. I thought maybe that could be source of the duplicates. I tried to use only CATEGORICAL features, that seemed to reduce number of duplicates, but did not eliminate them entirely. 

@xingyousong given previous attempts, I tried to pass "client_id" to the `suggest` and turned out it samples a different set of trials. can you suggest if the following idea is not too far fetched:

```python
population = []
client_id = 0

while len(population) < POPULATION_SIZE:
    needed = POPULATION_SIZE - len(population)
    sub_population = client.suggest(needed, client_id=client_id)
    client_id += 1
    for s in sub_population:
        if is_invalid(s):
            s.complete(...infeasible...)
            continue
        if is_duplicate(s):
            if have_results(s):
                s.complete(...previous results....)
            else:
                # store s and complete it once results available
            continue
        population.append(s)         
```

I will keep track of all the suggestions and make sure that they are all completed whenever I have results. But because some suggestions may be invalid I would need to get extra suggestions before proceeding with the population. How the "client_id" works internally? Can I use it for this purpose?

> Please never use the low-level designer API unless you absolutely know what you're doing.

I would very much like to learn. Greatly appreciate if you can give me a thought in this topic: https://github.com/google/vizier/discussions/1205

Thanks for the patience.
Kind regards

**Comment by xingyousong** [comment]
Re: `client_id`: Ah yeah, I realized we didn't make this explicit, sorry about that. Yes, using `client_id` is important in distributed tuning scenarios, to make sure suggestions aren't duplicated. See our new [doc](https://oss-vizier.readthedocs.io/en/latest/guides/user/distributed.html) on distributed Vizier.

Basically, `client_id` should represent `worker_id` if you're parallelizing computation. 

Thus `client_id+=1` should never be a line in your code; it's either:

1. Don't use `client_id` at all in `suggest()` if you're using a single worker
2. Use distinct `client_ids` for each distinct worker if you're parallelizing the evaluation workload.

Closing for now since I think this is roughly resolved.
